logstash-core/src/main/java/org/logstash/config/ir/compiler/Dataset.java

```diff
@@ -0,0 +1,438 @@
+ * batch of input data. For now this is intentionally not implemented since we want to clear
+ * the data stored in the Dataset tree as early as possible and before the output operations
+ * finish. Once the whole execution tree including the output operation is implemented in
+ * Java, this API can be adjusted as such.</p>
+ */
+public interface Dataset {
+
+    /**
+     * Compute the actual contents of the backing {@link RubyArray} and cache them.
+     * Repeated invocations will be effectively free.
+     * @param batch Input {@link JrubyEventExtLibrary.RubyEvent} received at the root
+     * of the execution
+     * @param flush True if flushing flushable nodes while traversing the execution
+     * @param shutdown True if this is the last call to this instance's compute method because
+     * the pipeline it belongs to is shut down
+     * @param options Flush Option hash passed to {@link RubyIntegration.Filter#flush(RubyHash)} 
+     * @return Computed {@link RubyArray} of {@link JrubyEventExtLibrary.RubyEvent}
+     */
+    Collection<JrubyEventExtLibrary.RubyEvent> compute(RubyIntegration.Batch batch,
+        boolean flush, boolean shutdown, RubyHash options);
+
+    /**
+     * Removes all data from the instance and all of its parents, making the instance ready for
+     * use with a new set of input data.
+     */
+    void clear();
+
+    /**
+     * Root {@link Dataset}s at the beginning of the execution tree that simply pass through
+     * the given set of {@link JrubyEventExtLibrary.RubyEvent} and have no state.
+     */
+    Collection<Dataset> ROOT_DATASETS = Collections.singleton(
+        new Dataset() {
+            @Override
+            public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+                final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+                final RubyHash options) {
+                return batch.collect();
+            }
+
+            @Override
+            public void clear() {
+            }
+        }
+    );
+
+    /**
+     * <p>{@link Dataset} that contains all {@link JrubyEventExtLibrary.RubyEvent} of all of
+     * its dependencies and is assumed to be at the end of an execution.</p>
+     * This dataset does not require an explicit call to {@link Dataset#clear()} since it will
+     * automatically {@code clear} all of its parents.
+     */
+    final class TerminalDataset implements Dataset {
+
+        /**
+         * Empty {@link Collection} returned by this class's
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)} implementation.
+         */
+        private static final Collection<JrubyEventExtLibrary.RubyEvent> EMPTY_RETURN =
+            Collections.emptyList();
+
+        /**
+         * Trivial {@link Dataset} that simply returns an empty collection of elements.
+         */
+        private static final Dataset TRIVIAL = new Dataset() {
```

Comment:
Is this object just used to avoid null checks on `Dataset.from()` return values? I may be misunderstanding otherwise, please advise on what this is for :)

I am also not sure what the word "trivial" is intended to mean. Is this maybe better named "EMPTY_DATASET" ?